packages/langium/src/lsp/document-highlighter.ts

```diff
@@ -49,14 +49,13 @@ export class DefaultDocumentHighlighter implements DocumentHighlighter {
         const targetAstNode = this.references.findDeclaration(selectedNode)?.element;
         if (targetAstNode) {
             const refs: Array<[CstNode, DocumentHighlightKind]> = [];
-            if (equalURI(getDocument(targetAstNode).uri, document.uri)) {
-                const nameNode = this.findNameNode(targetAstNode);
-                if (nameNode) {
-                    refs.push([nameNode, this.getHighlightKind(nameNode)]);
+            const includeDeclaration = equalURI(getDocument(targetAstNode).uri, document.uri);
+            const options = {onlyLocal: true, includeDeclaration};
+            this.references.findReferences(targetAstNode, options).forEach(ref => {
+                const leaf = findLeafNodeAtOffset(rootNode, ref.segment.offset);
+                if (leaf) {
+                    refs.push([leaf, this.getHighlightKind(leaf)]);
                 }
```

Comment:
Suggestion: Trying to find the leaf of the offset here seems a bit usual. The old API required that, but the new doesn't. Something like this is more appropriate here:

```suggestion
                refs.push([ref.segment.range, this.getHighlightKind(ref)]);
```

Note that you'll have to change some signatures for that. `getHighlightKind` should only be as simple as:

```ts
/**
 * Override this method to determine the highlight kind of the given reference.
 */
protected getHighlightKind(reference: ReferenceDescription): DocumentHighlightKind {
  return DocumentHighlightKind.Read;
}
```

Reply:
Seems like we don't need the `getHighlightKind` function anymore

Reply:
It's a template method, that allows downstream users to easily modify some behavior, even if the default behavior is just to return `DocumentHighlightKind.Read`. Removing that would also make it way harder to override this minor (but important) detail.